### PR TITLE
Add image-captioning tutorial

### PR DESCRIPTION
**Pull request by mfl28**
Last but not least, this adds the image captioning tutorial.

* The implementation is based on the [Show, Attend and Tell](https://arxiv.org/abs/1502.03044) paper (with some modifications).
* Includes implementations for [BLEU](https://en.wikipedia.org/wiki/BLEU) score, step learning rate scheduler, Gaussian Blur 2D transformer (to make attention masks look nicer) and caption dataset with custom collation class.
* Contrary to the other tutorials, I placed the validation/testing and sampling code into their own templatized functions inside a separate header file. This is because the main function was getting way too long :). 
* [Flickr8k](https://github.com/jbrownlee/Datasets/releases/tag/Flickr8k) dataset is used as it is relatively small (1GB). Same as for the other datasets, this dataset is downloaded via CMake.
* The Google Colab notebook is updated to run the new tutorial, including attention visualizations (using [Google Colab widgets](https://colab.research.google.com/notebooks/widgets.ipynb)) for predicted captions for a number of sample images randomly drawn from the test set.
* Uses [cxxopts](https://github.com/jarro2783/cxxopts) argument parser, allowing to optionally change most parameters of the tutorial from the command line (e.g. in the Google Colab notebook a larger batch size (128) can be used compared to default (32)).

### CODE REVIEW

**Comment by prabhuomkar** [comment]
Thanks @mfl28 , will take some time to review since there are many file changes :)
